File: dart-api-examples/spoken-language-identification/bin/spoken_language_identification.dart
// Copyright (c)  2024  Xiaomi Corporation
import 'dart:io';

import 'package:args/args.dart';
import 'package:sherpa_onnx/sherpa_onnx.dart' as sherpa_onnx;
import './init.dart';

void main(List<String> arguments) async {
  await initSherpaOnnx();

  final parser = ArgParser()
    ..addOption('encoder', help: 'Path to the whisper encoder model')
    ..addOption('decoder', help: 'Path to the whisper decoder model')
    ..addOption('tail-paddings', help: 'Tail paddings for the whisper model', defaultsTo: '0')
    ..addOption('wav', help: 'Path to test.wav for language identification')
    ..addFlag('help', abbr: 'h', help: 'Show this help message', negatable: false);

  final res = parser.parse(arguments);
  if (res['help'] as bool) {
    print(parser.usage);
    exit(0);
  }

  if (res['encoder'] == null || res['decoder'] == null || res['wav'] == null) {
    print(parser.usage);
    exit(1);
  }

  final encoder = res['encoder'] as String;
  final decoder = res['decoder'] as String;
  final tailPaddings = int.tryParse(res['tail-paddings'] as String) ?? 0;
  final wav = res['wav'] as String;

  final whisperConfig = sherpa_onnx.SpokenLanguageIdentificationWhisperConfig(
    encoder: encoder,
    decoder: decoder,
    tailPaddings: tailPaddings,
  );

  final config = sherpa_onnx.SpokenLanguageIdentificationConfig(
    whisper: whisperConfig,
    numThreads: 1,
    debug: true,
    provider: 'cpu',
  );

  final slid = sherpa_onnx.SpokenLanguageIdentification(config);

  final waveData = sherpa_onnx.readWave(wav);

  final stream = slid.createStream();
  stream.acceptWaveform(samples: waveData.samples, sampleRate: waveData.sampleRate);

  final result = slid.compute(stream);

  print('File: $wav');
  print('Detected language: ${result.lang}');

  stream.free();
  slid.free();
}

File: rust-api-examples/examples/spoken_language_identification.rs
// Copyright (c) 2026 Xiaomi Corporation
//
// This file demonstrates how to use sherpa-onnx's Rust API for spoken language
// identification.
//
// See ../README.md for how to run it.

use clap::Parser;
use sherpa_onnx::{
    SpokenLanguageIdentification, SpokenLanguageIdentificationConfig,
    SpokenLanguageIdentificationWhisperConfig, Wave,
};
use std::time::Instant;

#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
    #[arg(long)]
    wav: String,

    #[arg(long)]
    whisper_encoder: String,

    #[arg(long)]
    whisper_decoder: String,

    #[arg(long, default_value_t = 0)]
    tail_paddings: i32,

    #[arg(long, default_value_t = 1)]
    num_threads: i32,

    #[arg(long, default_value = "cpu")]
    provider: String,

    #[arg(long, default_value_t = false)]
    debug: bool,
}

fn main() -> anyhow::Result<()> {
    let args = Args::parse();
    let wave = Wave::read(&args.wav).ok_or_else(|| anyhow::anyhow!("Failed to read WAV file"))?;
    let audio_duration = wave.num_samples() as f64 / wave.sample_rate() as f64;

    let config = SpokenLanguageIdentificationConfig {
        whisper: SpokenLanguageIdentificationWhisperConfig {
            encoder: Some(args.whisper_encoder),
            decoder: Some(args.whisper_decoder),
            tail_paddings: args.tail_paddings,
        },
        num_threads: args.num_threads,
        provider: Some(args.provider),
        debug: args.debug,
    };

    let slid = SpokenLanguageIdentification::create(&config)
        .ok_or_else(|| anyhow::anyhow!("Failed to create SpokenLanguageIdentification"))?;

    let stream = slid.create_stream();
    let start = Instant::now();
    stream.accept_waveform(wave.sample_rate(), wave.samples());
    let result = slid
        .compute(&stream)
        .ok_or_else(|| anyhow::anyhow!("Failed to compute spoken language identification result"))?;
    let elapsed = start.elapsed().as_secs_f64();

    println!("File: {}", args.wav);
    println!("Detected language: {}", result.lang);
    println!("Elapsed seconds: {:.3}", elapsed);
    println!("Audio duration in seconds: {:.3}", audio_duration);
    println!("RTF: {:.3}/{:.3} = {:.3}", elapsed, audio_duration, elapsed / audio_duration);

    Ok(())
}
